Mock figure8 and aws functions for testing

ngreenwald opened the pull request:
This PR adds minimal tests for the Figure8 and AWS functions to try and catch syntax errors and other bugs that currently aren't ever tested without manually trying to upload and download files prior to submitting a PR. 

It also makes a few bugfixes and fixes typos as a result of some PRs not being merged in yet. Closes #103

willgraf commented on caliban_toolbox/aws_functions_test.py:
```diff
@@ -0,0 +1,76 @@
+import pathlib
+
+
+# TODO: What is a better way to mock the s3 = connect_aws() call within this function
+@patch('caliban_toolbox.aws_functions.connect_aws')
+def test_aws_upload_files(connect_aws):
```
I like to use the `pytest-mock` plugin for mocking. It includes a `mocker` pytest fixture which you can use like:

```python
class FakeS3(object):

    def __init__(self, *_, **__):
        pass

    def client(self, *_, **__):
        return self

    def upload_file(self, x1, x2, x3, Callback, ExtraArgs):
        return self

    def download_file(self, Bucket, Key, Filename):
        return self


def test_aws_upload_files(mocker):
    # have to use the caliban_toolbox path because it is used: `from getpass import getpass`
    mocker.patch('caliban_toolbox.aws_functions.getpass', lambda *x: None)
    # if instead we just use `import getpass` and call it like `getpass.getpass`
    # we could do the following:
    # mocker.patch('getpass.getpass', lambda *x: None)
    mocker.patch('boto3.Session', FakeS3)

    local_files = ['npz_file_' + str(num) for num in range(5)]
    aws_paths = ['aws_bucket/folder/npz_file_' + str(num) for num in range(5)]

    with tempfile.TemporaryDirectory() as temp_dir:
        for file in local_files:
            pathlib.Path(os.path.join(temp_dir, file)).touch()

        local_paths = [os.path.join(temp_dir, file) for file in local_files]

        aws_functions.aws_upload_files(local_paths=local_paths, aws_paths=aws_paths)
```

willgraf commented on caliban_toolbox/figure_eight_functions_test.py:
```diff
@@ -86,3 +96,105 @@ def test_create_job_urls():
+                                                       aws_folder='aws', stage='stage')
+
+
+def test_unzip_report():
+    with tempfile.TemporaryDirectory() as temp_dir:
+        # create example zip file
+        pathlib.Path(os.path.join(temp_dir, 'example_file.csv')).touch()
+        zip_path = os.path.join(temp_dir, 'job_report.zip')
+        zipfile.ZipFile(zip_path, mode='w').write(os.path.join(temp_dir, 'example_file.csv'))
+
+        figure_eight_functions.unzip_report(temp_dir)
+
+        assert os.path.exists(os.path.join(temp_dir, 'job_report.csv'))
```
Use the `tmpdir` pytest fixture instead.

```suggestion
def test_unzip_report(tmpdir):
    tmpdir = str(tmpdir)
    with tempfile.TemporaryDirectory() as temp_dir:
    # create example zip file
    # actually tmpdir starts as a Path object so instead of str(tmpdir) you can try to add them as Paths?
    pathlib.Path(os.path.join(temp_dir, 'example_file.csv')).touch()
    zip_path = os.path.join(temp_dir, 'job_report.zip')
    zipfile.ZipFile(zip_path, mode='w').write(os.path.join(temp_dir, 'example_file.csv'))

    figure_eight_functions.unzip_report(temp_dir)

    assert os.path.exists(os.path.join(temp_dir, 'job_report.csv'))
```

willgraf commented on caliban_toolbox/figure_eight_functions_test.py:
```diff
@@ -86,3 +96,105 @@ def test_create_job_urls():
+    download_report.return_value = 200
+    aws_download_files.return_value = 200
+
+    with tempfile.TemporaryDirectory() as temp_dir:
```
Same here, try using the `tmpdir` pytest fixture.

willgraf commented on requirements-test.txt:
```diff
@@ -3,3 +3,4 @@ pytest==4.6.5
 pytest-cov==2.5.1
 pytest-pep8>=1.0.6,<2.0.0
 coveralls>=1.8.2,<2.0.0
+requests_mock
```
As mentioned elsewhere, I'm a big fan of `pytest-mock` and the `mocker` fixture.

```suggestion
requests_mock
pytest-mock
```

willgraf commented on caliban_toolbox/figure_eight_functions_test.py:
```diff
@@ -86,3 +96,105 @@ def test_create_job_urls():
+
+    with tempfile.TemporaryDirectory() as temp_dir:
+
+        # create logs directory with zipped report
+        os.makedirs(os.path.join(temp_dir, 'logs'))
+        pathlib.Path(os.path.join(temp_dir, 'logs', 'example_file.csv')).touch()
+        zip_path = os.path.join(temp_dir, 'logs', 'job_report.zip')
+        zipfile.ZipFile(zip_path, mode='w').write(os.path.join(temp_dir, 'logs',
+                                                               'example_file.csv'))
+        # create log file
+        log_file = pd.DataFrame({'job_id': [1234]})
+        log_file.to_csv(os.path.join(temp_dir, 'logs', 'stage_0_upload_log.csv'))
+
+        figure_eight_functions.download_figure_eight_output(temp_dir)
```
Can we make any correct-ness assertions? Right now the code does not fail, which is great, but is it actually doing the right thing?

willgraf commented on caliban_toolbox/figure_eight_functions_test.py:
```diff
@@ -86,3 +96,105 @@ def test_create_job_urls():
+# testing of the different parts working together. However, better than not testing at all
+
+
+@patch('caliban_toolbox.figure_eight_functions.upload_log_file')
+@patch('caliban_toolbox.figure_eight_functions.aws_upload_files')
+@patch('caliban_toolbox.figure_eight_functions.copy_job')
+@patch("getpass.getpass")
```
I'd prefer to replace these with `mocker.patch` calls, but the difference in the quotations is bugging me.

```suggestion
@patch('caliban_toolbox.figure_eight_functions.upload_log_file')
@patch('caliban_toolbox.figure_eight_functions.aws_upload_files')
@patch('caliban_toolbox.figure_eight_functions.copy_job')
@patch('getpass.getpass')
```

willgraf commented on caliban_toolbox/figure_eight_functions_test.py:
```diff
@@ -86,3 +96,105 @@ def test_create_job_urls():
+@patch('caliban_toolbox.figure_eight_functions.aws_upload_files')
+@patch('caliban_toolbox.figure_eight_functions.copy_job')
+@patch("getpass.getpass")
+def test_create_figure_eight_job(getpass, copy_job, aws_upload_files, upload_log_file):
+    getpass.return_value = 'test_api_key'
+    copy_job.return_value = 123
+    aws_upload_files.return_value = '200'
+    upload_log_file.return_value = 567
+
+    with tempfile.TemporaryDirectory() as temp_dir:
+
+        # create crop directory
+        crop_dir = os.path.join(temp_dir, 'crop_dir')
+        os.makedirs(crop_dir)
+        np.savez(os.path.join(crop_dir, 'test_crop.npz'))
+
+        figure_eight_functions.create_figure_eight_job(base_dir=temp_dir, job_id_to_copy=123,
+                                                       aws_folder='aws', stage='stage')
```
`tmpdir` fixture to the rescue!

```suggestion
def test_create_figure_eight_job(getpass, copy_job, aws_upload_files, upload_log_file, tmpdir):
    tmpdir = str(tmpdir)
    getpass.return_value = 'test_api_key'
    copy_job.return_value = 123
    aws_upload_files.return_value = '200'
    upload_log_file.return_value = 567


    # create crop directory
    crop_dir = os.path.join(temp_dir, 'crop_dir')
    os.makedirs(crop_dir)
    np.savez(os.path.join(crop_dir, 'test_crop.npz'))

    figure_eight_functions.create_figure_eight_job(base_dir=temp_dir, job_id_to_copy=123,
                                                   aws_folder='aws', stage='stage')
```

willgraf commented on caliban_toolbox/figure_eight_functions_test.py:
```diff
@@ -86,3 +96,105 @@ def test_create_job_urls():
+        url_encoded_dict = urllib.parse.urlencode(url_dict)
+        url = url.format(test_job_id, url_encoded_dict)
+
+        # mock the call
+        response_dict = {'status_code': 200}
+        m.put(url, text=json.dumps(response_dict))
+        figure_eight_functions.upload_log_file(log_file=example_log_string, job_id=test_job_id,
+                                               key=test_key)
```
Can we make any assertions about the function? If not, maybe it should return something to indicate success or failure?

willgraf commented on caliban_toolbox/figure_eight_functions_test.py:
```diff
@@ -86,3 +96,105 @@ def test_create_job_urls():
     #                                                               rgb_mode=rgb_mode)
+
+
+# TODO: Is this test useful?
+def test_copy_job():
+    with requests_mock.Mocker() as m:
+
+        # create test data
+        test_job_id = 666
+        test_appen_key = 'a1b2c3'
+        return_id = 123
+        return_dict = {'status_code': 200, 'id': return_id}
+
+        # generate same url as function for mocking
+        url = 'https://api.appen.com/v1/jobs/{}/copy.json?'.format(str(test_job_id))
+
+        # mock the call
+        m.get(url, text=json.dumps(return_dict))
+        new_job_id = figure_eight_functions.copy_job(job_id=test_job_id, key=test_appen_key)
+
+        assert new_job_id == return_id
```
It does assert that we are extracting the `'id'` field from the response JSON.

It could be more useful if you also test an error/failure.

ngreenwald commented:
Thanks for the great pointers! Here's what I changed:

- Replaced all of the @patch decorators were mocker.patch calls within the functions
- Switched all tempfile.TemporaryDirectory() context managers to tmp_path pytest decorators
- Wherever possible, mocked the actual call within a helper function that was causing the issue, rather than the entire helper function itself, to increase test coverage of interactions between different functions.

willgraf commented on caliban_toolbox/aws_functions_test.py:
```diff
@@ -0,0 +1,72 @@
+        pathlib.Path(Filename).touch()
+
+
+def test_aws_upload_files(mocker, tmp_path):
```
Oh nice, I was unaware of `tmp_path` and have only used `tmpdir`.

willgraf commented on caliban_toolbox/aws_functions_test.py:
```diff
@@ -0,0 +1,72 @@
+    for file in local_files:
+        pathlib.Path(os.path.join(tmp_path, file)).touch()
+
+    local_paths = [os.path.join(tmp_path, file) for file in local_files]
+
+    aws_functions.aws_upload_files(local_paths=local_paths, aws_paths=aws_paths)
```
There's also a concept of [a `spy`](https://github.com/pytest-dev/pytest-mock/#spy) which watches other functions, and you can assert they were called with a certain value.

Maybe we can `spy` on spy on `s3.upload_file` to make sure its being called with what we expect? I'm not sure how useful this is, just an idea.

ngreenwald commented:
I added the option to raise an error within the FakeS3 object so that we can test out the error handling code. However, when I try and initialize the object with the right params so that it will generate those errors, I get an object not callable error. 

`mocker.patch('boto3.Session', FakeS3(raise_error='missing'))`


I think this is because we're passing in the class itself beforehand, whereas now I've actually created an instantiation of the class? Not sure:

``` 
session = boto3.Session(aws_access_key_id=AWS_ACCESS_KEY_ID,
                              aws_secret_access_key=AWS_SECRET_ACCESS_KEY)
E       TypeError: 'FakeS3' object is not callable
```

This is for the test_aws_download_files() function

willgraf commented:
Yep you nailed it. Try wrapping it in a lambda function:

```python
mocker.patch('boto3.Session', lambda: FakeS3(raise_error='missing'))
```

ngreenwald commented:
I added in more robust tests for the FakeS3 object. This allows us to test out the error handling code in the aws_download function, and also that the missing file names are properly propagated up to the main function. 

I added checks for the other functions as well that were missing basic assert statements. 
I also added the FakeS3 object to the Figure8 testing functions to dive into the aws function calls.
